File: internal/notes/position.go
package notes

// Sibling ordering uses plain float64 positions ("fractional positioning"):
// inserting between two siblings takes the arithmetic midpoint, and
// appending takes the last position plus a fixed gap. This gives the same
// practical benefits as string-based fractional indexing (reorder without
// rewriting siblings, clients can compute a valid position offline without
// coordinating with the server) with much simpler, easily-verified
// arithmetic. float64 has 52 bits of mantissa, i.e. room for dozens of
// successive same-spot inserts before precision would become a concern —
// far beyond what a personal note hierarchy will ever need. The frontend
// mirrors this exact logic in TypeScript (see frontend/src/lib/position.ts).

const gap = 1000.0

// PositionAfter returns a position placed after the last known sibling
// position (or the first position if the list is empty).
func PositionAfter(last *float64) float64 {
	if last == nil {
		return gap
	}
	return *last + gap
}

// PositionBefore returns a position placed before the first known sibling
// position (or the first position if the list is empty).
func PositionBefore(first *float64) float64 {
	if first == nil {
		return gap
	}
	return *first / 2
}

// PositionBetween returns a position strictly between two sibling
// positions. Callers must ensure before < after.
func PositionBetween(before, after float64) float64 {
	return before + (after-before)/2
}

File: frontend/src/lib/position.ts
// Mirrors internal/notes/position.go: sibling order is a plain float64
// ("fractional positioning" via arithmetic midpoints/gaps rather than
// string-based fractional indexing) so the client can compute a valid
// position for a drag&drop reorder or offline-created node without asking
// the server.
const GAP = 1000

export function positionAfter(last: number | null | undefined): number {
  if (last == null) return GAP
  return last + GAP
}

export function positionBefore(first: number | null | undefined): number {
  if (first == null) return GAP
  return first / 2
}

export function positionBetween(before: number, after: number): number {
  return before + (after - before) / 2
}
